Enable System Attributes and their use in LocationFilters

tehhowch opened the pull request:
Refs #3143 #526 

LocationFilter will now union a planet and its system, and a system and its planets, while checking for matches to a particular set of attributes.
A system's attributes are comprised of those specified via file, with the `attributes` node:
```
system "Sol Arach"
    attributes "nexus"
```
and those that are automatically generated from its StellarObjects and links.  The above "nexus" attribute would be unnecessary, as Sol Arach has 7 systems linked to it, and the "nexus" attribute is automatically assigned for those systems that have 6 or more linked systems.

Current automatic system attributes ( #3143 )
 - nexus
 - crossroads
 - galactic phenomena
 - ternary star
 - binary star
 - no star
 - high solar wind
 - high luminosity
 - high minable density
 - high asteroid density
 - many moons
 - wormhole (when unioned with `PlanetAttributes()` in LocationFilter::Matches, if the wormhole is accessible)
 - inhabited (when unioned with `PlanetAttributes()` in LocationFilter::Matches, if the inhabited planets are accessible)
 - uninhabited

The import of attributes from a system's accessible planets means the system will import any geographic or civilization based attributes (e.g. `pirate`, `north`, `dirt belt`, etc). It also means "contradictory" attributes can exist - for example, peacekeeping forces might need an escort to a system that is `urban rich poor` (note that because an individual planet won't have both attributes, contradictory attributes won't match if `and`ed in a source, destination, or stopover filter (but will work in a waypoint or NPC system filter).

An alternative (or perhaps complementary) approach would be to specify another embedded LocationFilter that is applied only during system match-checking.

tehhowch commented:
Updated the moon criteria based on game data posted in #3143

endless-sky commented:
Having that many auto-generated attributes doesn't really make sense to me. The auto-attributes for planets (spaceport, shipyard, and outfitter) have really obvious meanings and are likely to be useful, but many of these attributes have unclear meanings or limited usefulness.

tehhowch commented:
I felt it better to err on the side of exposing more keywords for mission locations than fewer, as it allows a bit more specificity when writing random missions and wanting to have appropriately descriptive text. They are definitely on the arbitrary side, but make it possible to write missions to escort ships that want to go to a mining-heavy system/asteroid-heavy system, smuggle a passenger from/to a system having many links, lead researchers to systems with crazy/multiple stars, lots of moons, etc, without needing to write one mission for each possible destination.

tehhowch commented:
I've dropped two auto-attributes in favor of two new ones based on 317b495a7217784bb490637564e1ff57e22c0193.
I also removed the static definitions of what constitutes a "galactic phenomena" into the `data/stars.txt` file, allowing plugin creators to easily change which sprites generate the "galactic phenomena" auto-attribute.
There were also some legacy includes in GameData so I've pruned them away.

endless-sky commented:
This still seems like way too many auto-attributes and other new features - again, considering the fact that once they're added, the game needs to support them all in perpetuity. It makes more sense to me to add the bare minimum that's clearly useful, and only expand it over time if there's a definite need.

In commit e1c1a55 I've added support for system attributes, with a single auto-attribute (uninhabited). That's the one attribute that's clearly going to be useful.
